---
title: Bayesian approaches to Transcription Factor Target Identification
venue: 'Course in Practical Systems Biology: Visualisation and Reverse engineering
  gene regulatory networks, EuroMediterranean University Centre of Ronzano, Bologna,
  Italy'
linkpdf: ftp://ftp.dcs.shef.ac.uk/home/neil/gp_bologna10.pdf
year: '2010'
month: 10
day: '10'
abstract: |-
  A simple approach to target identification through gene expression studies has been to cluster the expression profiles and look for coregulated genes within clusters. Within systems biology mechanistic models of gene expression are typically constructed through differential equations. mRNAâ€™s production is taken to be proportional to transcription factor activity (with the proportionality given by the sensitivity) and the mRNA is assumed to decay at a particular rate. The assumption that coregulated genes have similar profiles is equivalent to assuming both the decay and the sensitivity are high.\
  \
  In this lecture we introduce Bayesian approaches to target identification which make use of sampling approaches to rank candidate lists of targets. We will begin with an introduction to the target identification problem and an overview of the power of Bayesian approaches in solving it. We will then consider how probabilistic models such as Gaussian processes can be used for ranking potential targets of a transcription factor. These models are simple enough to allow genome wide target identification, but rich enough to encode dynamical behavior that, allowing us to identify putative targets even when decay rates are low.
layout: talk
key: Lawrence:eurogene10
categories:
- Lawrence:eurogene10
authors:
- firstname: Neil D.
  lastname: Lawrence
- firstname: Antti
  lastname: Honkela
published: 2010-10-10
---
